src/modules/Extension/Service.php

```diff
@@ -634,9 +647,89 @@ public function getCoreAndActiveModules()
         $extensionMod = $this->di['mod']('extension');
         $mods = $extensionMod->getCoreModules();
 
-        $result = array_merge($mods, $list);
-        sort($result);
+        $modules = array_merge($mods, $list);
+        sort($modules);
 
-        return $result;
+        return $modules;
+    }
+
+    public function getCoreAndActiveModulesAndPermissions(): array
+    {
+        $enabledModules = $this->getCoreAndActiveModules();
+        $modules = [];
+
+        foreach ($enabledModules as $module) {
+            $class = 'Box\Mod\\' . ucfirst($module) . '\Service';
+            if (class_exists($class) && method_exists($class, 'getModulePermissions')) {
+                $moduleService = new $class;
+                $moduleService->setDi($this->di);
+                $permissions = $moduleService->getModulePermissions();
+
+                if (isset($permissions['hide_permissions']) && $permissions['hide_permissions']) {
+                    continue;
+                } else {
+                    unset($permissions['hide_permissions']);
+                    $modules[$module]['permissions'] = $permissions;
+                }
+            } elseif ($module !== 'index' && $module !== 'dashboard') {
+                $modules[$module]['permissions'] = [];
+            }
+        }
+
+        return $modules;
+    }
+
+    public function getSpecificModulePermissions(string $module): array
+    {
+        $class = 'Box\Mod\\' . ucfirst($module) . '\Service';
+        if (class_exists($class) && method_exists($class, 'getModulePermissions')) {
+            $moduleService = new $class;
+            $moduleService->setDi($this->di);
+            $permissions = $moduleService->getModulePermissions();
+
+            if (isset($permissions['hide_permissions']) && $permissions['hide_permissions']) {
+                return [];
+            } else {
+                unset($permissions['hide_permissions']);
+                return $permissions;
+            }
+        }
+        return [];
+    }
+
+    // Checks if the current user has permission to edit a module's settings
+    public function checkPermission(string $module, \Box_App|null $app = null): void
+    {
+        $staff_service = $this->di['mod_service']('Staff');
+        $modules = $this->getCoreAndActiveModulesAndPermissions();
```

Comment:
This is likely going to be a performance impact when there are a lot of modules with a lot of permissions. I imagine that checkPermission will be called a lot, and calling getCoreAndActiveModulesAndPermissions() each time sounds like a lot of compute time for each checkPermission call.

Reply:
In this case, not really.
This check is specific to if someone is accessing the settings of a module (`/admin/extension/settings/email` for example), so it shouldn't be happening all that frequently.

Reply:
Optimized it :)